Implement support for the StringContext IDL attribute in code generator

lukewarlow opened the pull request:
#### f6ac260611c27959abdcb4ab2f258ef40dee99a7
<pre>
Implement support for the StringContext IDL attribute in code generator
<a href="https://bugs.webkit.org/show_bug.cgi?id=267581">https://bugs.webkit.org/show_bug.cgi?id=267581</a>

Reviewed by NOBODY (OOPS!).

This adds support for the StringContext IDL attribute from the trusted types spec to the code generator.
No CSP enforcement is implemented as part of this change.

Spec: <a href="https://w3c.github.io/trusted-types/dist/spec/#StringContext">https://w3c.github.io/trusted-types/dist/spec/#StringContext</a>

* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
(WebCore::TrustedTypeVisitor::operator()):
(WebCore::processValueWithDefaultPolicy):
(WebCore::getTrustedTypeCompliantString):
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
(WebCore::Converter&lt;IDLStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLAtomStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLAtomStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLAtomStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateParametersCheck):
(IsAnnotatedType):
(GetAnnotatedIDLType):
(JSValueToNative):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestStringContextConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.cpp: Added.
(WebCore::JSTestStringContextDOMConstructor::prototypeForStructure):
(WebCore::JSTestStringContextDOMConstructor::initializeProperties):
(WebCore::JSTestStringContextPrototype::finishCreation):
(WebCore::JSTestStringContext::JSTestStringContext):
(WebCore::JSTestStringContext::createPrototype):
(WebCore::JSTestStringContext::prototype):
(WebCore::JSTestStringContext::getConstructor):
(WebCore::JSTestStringContext::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringBody):
(WebCore::JSTestStringContext::subspaceForImpl):
(WebCore::JSTestStringContext::analyzeHeap):
(WebCore::JSTestStringContextOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestStringContextOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestStringContext::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.h: Added.
(WebCore::JSTestStringContext::create):
(WebCore::JSTestStringContext::createStructure):
(WebCore::JSTestStringContext::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestStringContext.idl: Added.
* Source/WebCore/dom/TrustedTypePolicy.cpp:
(WebCore::TrustedTypePolicy::createHTML):
(WebCore::TrustedTypePolicy::createScript):
(WebCore::TrustedTypePolicy::createScriptURL):
(WebCore::TrustedTypePolicy::getPolicyValue):
* Source/WebCore/dom/TrustedTypePolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::requireTrustedTypesForSinkGroup const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6ac260611c27959abdcb4ab2f258ef40dee99a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38316 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17259 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40640 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40874 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34077 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20022 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14596 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32310 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14566 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12675 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34242 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42152 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38496 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13241 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36694 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14799 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13649 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14263 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -170,6 +176,195 @@ template<typename T> struct JSConverter<IDLLegacyNullToEmptyStringAdaptor<T>> {
+    static JSC::JSValue convert(JSC::JSGlobalObject& lexicalGlobalObject, const AtomString& value)
+    {
+        return JSConverter<T>::convert(lexicalGlobalObject, value);
+    }
```
TODO: Add the below and update the IDL test to include a reflected URL Trusted type attribute to check it compiles properly.
```
    static JSC::JSValue convert(JSC::JSGlobalObject& lexicalGlobalObject, const URL& value)
    {
        return JSConverter<T>::convert(lexicalGlobalObject, value.string());
    }
    ```

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    if (auto* trustedHTML = JSTrustedHTML::toWrapped(vm, value))
+        return trustedHTML->toString();
+
+    // TODO: Implement enforcement logic
```
Comments should end with a period as per WebKit coding style.

darinadler replied:
Also, we use `FIXME`, not `TODO`.

darinadler replied:
Is the enforcement particularly complicated? I was wondering why we don’t have a first cut at the enforcement logic in this initial patch.

lukewarlow replied:
Yeah the enforcement involves checking CSP and calling into default policies in specific cases. I could do it in one go but previously was told to keep IDL changes separate to keep PRs smaller.

darinadler replied:
This seems to do a bit too little; like at least we should have the enforcement function even if it has a placeholder simplistic rule, instead of many different `TODO` comments.

lukewarlow replied:
Okay I'll do some refactoring of this so it's a bit clearer what the final steps might look like even if the implementation itself isn't there (CSP parsing isn't merged yet else I could do a bit more).

lukewarlow replied:
I've added some of the structure of the enforcement but it will require CSP changes to implement fully (have a follow up for that)

lukewarlow replied:
@darinadler the CSP parsing is merged now so I could go further and actually include the CSP enforcement code too if it's preferable for this to be done in one hit? Unless I also apply the [StringContext] IDL attribute to some of the sinks in this PR the CSP enforcement won't have test coverage though?

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
     return replaceUnpairedSurrogatesWithReplacementCharacter(WTFMove(string));
 }
 
+String valueToTrustedHTMLString(JSGlobalObject& lexicalGlobalObject, JSValue value, const String& sink, bool nullToEmpty)
```
`nullToEmpty` is not a great name IMO. Maybe `shouldConvertNullToEmptyString`?

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    return nullString();
+}
+
+AtomString valueToTrustedHTMLAtomString(JSGlobalObject& lexicalGlobalObject, JSValue value, const String& sink, bool nullToEmpty)
```
ditto.

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    return nullAtom();
+}
+
+String valueToTrustedScriptString(JSGlobalObject& lexicalGlobalObject, JSValue value, const String& sink, bool nullToEmpty)
```
ditto.

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    return nullString();
+}
+
+AtomString valueToTrustedScriptAtomString(JSGlobalObject& lexicalGlobalObject, JSValue value, const String& sink, bool nullToEmpty)
```
ditto.

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    if (auto* trustedScript = JSTrustedScript::toWrapped(vm, value))
+        return trustedScript->toString().toExistingAtomString();
+
+    // TODO: Implement enforcement logic
```
ditto.

cdumez replied:
Please apply these 2 to the rest of the file.

cdumez commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -170,6 +176,255 @@ template<typename T> struct JSConverter<IDLLegacyNullToEmptyStringAdaptor<T>> {
+
+    static String convert(JSC::JSGlobalObject& lexicalGlobalObject, JSC::JSValue value, const String& sink)
+    {
+        auto result = valueToTrustedHTMLString(lexicalGlobalObject, value, sink, false);
```
We don't pass booleans like this as parameter because it is too opaque at call sites. We have no idea what `false` means here. We should switch to an enum class:
```
enum class ShouldConvertNullToEmptyString : bool { No, Yes };
```

Then pass `ShouldConvertNullToEmptyString::No` here.

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -7412,6 +7424,19 @@ sub JSValueToNative
     push(@conversionArguments, $globalObjectReference) if JSValueToNativeDOMConvertNeedsGlobalObject($type);
     push(@conversionArguments, $exceptionThrower) if $exceptionThrower;
+    if ($type->extendedAttributes->{StringContext})
+    {
```
You should check the rest of the file, I doubt putting the curly brackets like this matches our coding style. I bet the curly bracket should be at the end of the previous line.

lukewarlow replied:
Apologies I must have looked at the function style and copied that in this case. Will fix.

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -7412,6 +7424,19 @@ sub JSValueToNative
+    {
+        my $interfaceName = $interface->type->name;
+        if ($functionName)
+        {
```
ditto.

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -7412,6 +7424,19 @@ sub JSValueToNative
+        {
+            push(@conversionArguments, "\"$interfaceName $functionName\"_s");
+        }
+        else
```
ditto.

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -7412,6 +7424,19 @@ sub JSValueToNative
+            push(@conversionArguments, "\"$interfaceName $functionName\"_s");
+        }
+        else
+        {
```
ditto.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
 
+String valueToTrustedHTMLString(JSGlobalObject& lexicalGlobalObject, JSValue value, const String& sink, bool nullToEmpty)
+{
+    UNUSED_PARAM(sink);
```
Style-wise we prefer to comment out, or omit, the `sink` name in the arguments list above. `UNUSED_PARAM` should only be used when we can’t do that for some reason.

lukewarlow replied:
Ah okay good to know. Will change that.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    VM& vm = lexicalGlobalObject.vm();
+
+    if (auto* trustedHTML = JSTrustedHTML::toWrapped(vm, value))
+        return trustedHTML->toString();
```
We need to find a way to cut down on the boilerplate and repetitiveness of these functions. All these different ones and this line of code is the only difference. There has to be some technique where we don’t just paste the code over and over again. The simplest way might be to have a helper function that takes a lambda for converting the HTML object to a string, or there could be some other technique.

lukewarlow replied:
In the final run through it will be split into a separate function so there's far less repetition. I'm happy to make changes to this initial patch to achieve a similar effect.

lukewarlow replied:
I've refactored this to more closely follow the spec shape and as a result this is much more streamlined.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,94 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    VM& vm = lexicalGlobalObject.vm();
+
+    if (auto* trustedHTML = JSTrustedHTML::toWrapped(vm, value))
+        return trustedHTML->toString().toExistingAtomString();
```
I don’t understand why the `AtomString` is guaranteed to exist here. Not sure that `toString().toExistingAtomString()` is the correct idiom for what we are trying to do.

lukewarlow replied:
Ah apologies I forgot about this I meant to check what the correct way to convert from a string to an Atom String was.

lukewarlow replied:
Given a String what is the correct way to convert it to an AtomString?

cdumez replied:
`AtomString { string }` simply

lukewarlow replied:
Thanks don't know how I missed that

lukewarlow committed:
Implement support for the StringContext IDL attribute in code generator
https://bugs.webkit.org/show_bug.cgi?id=267581

Reviewed by NOBODY (OOPS!).

This adds support for the StringContext IDL attribute from the trusted types spec to the code generator.
No CSP enforcement is implemented as part of this change.

Spec: https://w3c.github.io/trusted-types/dist/spec/#StringContext

* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
(WebCore::TrustedTypeVisitor::operator()):
(WebCore::processValueWithDefaultPolicy):
(WebCore::getTrustedTypeCompliantString):
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
(WebCore::Converter<IDLStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::Converter<IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::Converter<IDLStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::JSConverter<IDLStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::Converter<IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::Converter<IDLStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::Converter<IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::Converter<IDLAtomStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::Converter<IDLAtomStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::Converter<IDLAtomStringStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringStringContextTrustedScriptURLAdaptor<T>>::convert):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateParametersCheck):
(IsAnnotatedType):
(GetAnnotatedIDLType):
(JSValueToNative):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestStringContextConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.cpp: Added.
(WebCore::JSTestStringContextDOMConstructor::prototypeForStructure):
(WebCore::JSTestStringContextDOMConstructor::initializeProperties):
(WebCore::JSTestStringContextPrototype::finishCreation):
(WebCore::JSTestStringContext::JSTestStringContext):
(WebCore::JSTestStringContext::createPrototype):
(WebCore::JSTestStringContext::prototype):
(WebCore::JSTestStringContext::getConstructor):
(WebCore::JSTestStringContext::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringBody):
(WebCore::JSTestStringContext::subspaceForImpl):
(WebCore::JSTestStringContext::analyzeHeap):
(WebCore::JSTestStringContextOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestStringContextOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestStringContext::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.h: Added.
(WebCore::JSTestStringContext::create):
(WebCore::JSTestStringContext::createStructure):
(WebCore::JSTestStringContext::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestStringContext.idl: Added.
* Source/WebCore/dom/TrustedTypePolicy.cpp:
(WebCore::TrustedTypePolicy::createHTML):
(WebCore::TrustedTypePolicy::createScript):
(WebCore::TrustedTypePolicy::createScriptURL):
(WebCore::TrustedTypePolicy::getPolicyValue):
* Source/WebCore/dom/TrustedTypePolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::requireTrustedTypesForSinkGroup const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +147,104 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
 }
 
+// https://w3c.github.io/trusted-types/dist/spec/#process-value-with-a-default-policy-algorithm
+static inline std::variant<std::nullptr_t, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> processValueWithDefaultPolicy(VM& vm, ScriptExecutionContext& scriptExecutionContext, const String& expectedType, const String& input, const String& sink)
```
This is a static inline rather than being declared in the header because I can't import the Trusted types headers into the header file and std::variant can't make use of forward declarations. In future this function will probably end up in a separate helper file as it will be used in more places.

rniwa commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -22,8 +22,18 @@
+#include "WorkerGlobalScope.h"
 #include <JavaScriptCore/HeapInlines.h>
+#include <JavaScriptCore/JSCast.h>
+#include <JavaScriptCore/JSCInlines.h>
```
JSInline should appear before JSCast since uppercase letters appear before lowercase letters.

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -31,13 +31,17 @@
 WEBCORE_EXPORT String valueToUSVString(JSC::JSGlobalObject&, JSC::JSValue);
 WEBCORE_EXPORT AtomString valueToUSVAtomString(JSC::JSGlobalObject&, JSC::JSValue);
+String getTrustedTypeCompliantString(const String& expectedType, JSC::JSGlobalObject&, JSC::JSValue, const String& sink, ShouldConvertNullToEmptyString);
+WEBCORE_EXPORT String getTrustedTypeCompliantString(const String& expectedType, ScriptExecutionContext&, const String& value, const String& sink);
```
Extracted to a separate function so it can be used elsewhere in future PRs
